presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ImplementLimitWithTiesOverSort.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.metadata.FunctionKind;
+import io.prestosql.metadata.Signature;
+import io.prestosql.spi.type.StandardTypes;
+import io.prestosql.spi.type.TypeSignature;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.LimitNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.SortNode;
+import io.prestosql.sql.planner.plan.WindowNode;
+import io.prestosql.sql.tree.ComparisonExpression;
+import io.prestosql.sql.tree.FrameBound;
+import io.prestosql.sql.tree.FunctionCall;
+import io.prestosql.sql.tree.GenericLiteral;
+import io.prestosql.sql.tree.QualifiedName;
+import io.prestosql.sql.tree.Window;
+import io.prestosql.sql.tree.WindowFrame;
+
+import java.util.Optional;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.sql.planner.plan.Patterns.limit;
+import static io.prestosql.sql.planner.plan.Patterns.sort;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+
+/**
+ * Transforms:
+ * <pre>
+ * - Limit (row count = x, withTies = true)
+ *    - Sort (order by a,b,c)
+ * </pre>
+ * Into:
+ * <pre>
+ * - Project (prune rank symbol)
+ *    - Filter (rank <= x)
+ *       - Window (function: rank, order by a,b,c)
+ *          - Sort (order by a,b,c)
+ * </pre>
+ */
+public class ImplementLimitWithTiesOverSort
```

Comment:
Seeing this "non-trivial" rewrite that looks almost the same code as the other ones got me thinking about whether we're representing "with ties" properly in the plan. Let me describe my thought process and then suggest how we might want to do it:

1. The rewrites look similar and are not trivial so we should somehow share the implementation
2. At a fundamental level, what we're doing is look for a "limit with ties" over something that is "ordered"
3. One option would be to somehow tag the Sort / TopN node with a marker interface and a single method to get the ordering they compute. However, this wouldn't work for the case with project in between -- we'd still need separate rules.
4. Could we model this with traits (not supported yet, but...)? Unfortunately, this is not sufficient. We could attempt to match a "limit with ties" over "something that guarantees some kind of ordering". Traits might describe a tighter guarantee than what was originally in the ORDER BY clause. Imagine, for instance, a case where the data source is sorted by more columns than show up in the ORDER BY clause.

Which leads me to the following conclusion: "limit with ties" is not just a simple limit and a flag. *The set of columns used to define the ties is meaningful*. Conceptually one could think of `FETCH FIRST` vs `ORDER BY ... FETCH FIRST WITH TIES` as two completely different operations that happen to share some of the same syntax in the language.

And the suggestion on how we might approach it: add the ordering scheme to "limit with ties" (or create a new node for it). These rules would just find a "limit with ties" and rewrite into the new form without worrying about whether there's a sort or topN or anything else under them.

Reply:
> Which leads me to the following conclusion: "limit with ties" is not just a simple limit and a flag. The set of columns used to define the ties is meaningful.
Of course, `FETCH FIRST ... WITH TIES` must be aware of the exact ordering scheme from the corresponding `ORDER BY` clause. Adding the ordering scheme to the LimitNode would simplify things a lot.

However, that would be quite a different approach from how `ORDER BY ... LIMIT` is handled.
As I recall, LimitNode and SortNode are planned distinctly, possibly with an OffsetNode between them, and there are rules responsible for merging them into TopNNode.
I thought it was right to let QueryPlanner simply _rewrite_  a query into a plan and let the rules take care of the implementation.

Reply:
Maybe the analogy between 'with ties' and TopN isn't so close. Pairing a LimitNode with ordering scheme to create a TopNNode is a matter of _optimisation_, while in the case of 'with ties' it is a necessity.

Reply:
Another thought.
If we are going to add the ordering scheme to LimitNode with ties, then why don't we add it to LimitNode without ties, too? I'm aware this idea doesn't comply with the principle of QueryPlanner being merely a translator AST -> plan. However, there are benefits:
- simplicity at planning stage (no check for with ties)
- easy transformation of LimitNode without ties into TopNNode (currently there are multiple rules to check if the source is sorted / project + sorted). Same as for LimitNode with ties.
- maybe we could even quit planning Sort _at all_ when Limit is present (LimitNode without ties would be transformed into TopNNode, and LimitNode with ties would be transformed into Window with OrderingScheme, which would subsequently emit a distributed Sort for efficiency (have you heard of this brilliant idea? ;) )